core/src/main/java/org/lflang/generator/docker/DockerGenerator.java

```diff
@@ -24,31 +32,174 @@ public DockerGenerator(LFGeneratorContext context) {
+    return defaultBuildCommands();
+  }
+
+  /** Return the command that sources the pre-build script, if there is one. */
+  protected List<String> getPreBuildCommand() {
+    var script = context.getTargetConfig().get(DockerProperty.INSTANCE).preBuildScript();
+    if (!script.isEmpty()) {
+      return List.of("source src-gen/" + script);
+    }
+    return List.of();
+  }
+
+  /** Return the command that sources the post-build script, if there is one. */
+  protected List<String> getPostBuildCommand() {
+    var script = context.getTargetConfig().get(DockerProperty.INSTANCE).postBuildScript();
+    if (!script.isEmpty()) {
+      return List.of("source src-gen/" + script);
+    }
+    return List.of();
+  }
+
+  /** Generate a header to print at the top of the Dockerfile. */
+  protected String generateHeader() {
+    return """
+           # Generated by the Lingua Franca compiler version %s
+           #   - Docs: https://www.lf-lang.org/docs/handbook/containerized-execution"
+           """
+        .formatted(LocalStrings.VERSION);
+  }
+
+  /** Return the Docker RUN command used for building. */
+  protected String generateRunForBuild() {
+    return "RUN "
+        + StringUtil.joinObjects(
+            Stream.of(
+                    List.of("set -ex"),
+                    getPreBuildCommand(),
+                    getBuildCommands(),
+                    getPostBuildCommand())
+                .flatMap(java.util.Collection::stream)
+                .collect(Collectors.toList()),
+            " \\\n\t&& ");
+  }
+
+  /** Return the ENTRYPOINT command. */
+  protected String generateEntryPoint() {
+    return "ENTRYPOINT ["
+        + getEntryPointCommands().stream()
+            .map(cmd -> "\"" + cmd + "\"")
+            .collect(Collectors.joining(","))
+        + "]";
+  }
+
+  /** Return a COPY command to copy the executable from the builder to the runner. */
+  protected String generateCopyOfExecutable() {
+    var lfModuleName = context.getFileConfig().name;
+    return "COPY --from=builder /lingua-franca/%s/bin/%s ./bin/%s"
+        .formatted(lfModuleName, lfModuleName, lfModuleName);
   }
 
-  public String userRunCommand() {
-    var runCmd = context.getTargetConfig().get(DockerProperty.INSTANCE).run();
-    if (runCmd != null && !runCmd.isEmpty()) {
-      return "RUN " + runCmd;
+  /** Return a COPY command to copy the scripts from the builder to the runner. */
+  protected String generateCopyOfScript() {
+    var script = context.getTargetConfig().get(DockerProperty.INSTANCE).preRunScript();
+    if (!script.isEmpty()) {
+      return "COPY --from=builder /lingua-franca/%s/src-gen/%s ./scripts/%s"
+          .formatted(context.getFileConfig().name, script, script);
+    }
+    return "# (No pre-run script provided.)";
+  }
+
+  /**
+   * Return a list of strings used to construct and entrypoint. If this is done for a federate, then
+   * also include additional parameters to pass in the federate ID.
```

Comment:
```suggestion
   * also include additional parameters to pass in the federation ID.
```

At first I was surprised that the federation ID is hardcoded to 1, but now I see that it actually makes sense because use of distinct Docker networks serves the same purpose as using distinct federation IDs. I'm happy with this implementation choice.